Pre-RFC describing mechanism to remove orphan rule constraints (#3482)
rust-lang/rfcs: text/0000-extern-impl.md

unaffected by the `DBMongo` and `DBRocks` implementations.
```mermaid
graph TB
Bin(MysqlUser) --> DBMysql
DBMysql & DBMongo & DBRocks -->|impl| DBTrait
DBMysql --> Mysql
DBMongo --> MongoDB
DBRocks --> RocksDB
```
# Guide-level explanation (cargo)
[guide-level-explanation]: #guide-level-explanation 

Within a workspace, you can designate an "impl" relationship between two crates.
For example, in your `my_impls/Cargo.toml` you can put:

```
[dependency]
my_definitions = { version = "0.1", path = "../my_definitions", impl = True }
```

This means that your `my_impls` crate can add trait implementations using types
or traits defined in `my_definitions`, almost as if they were defined in
`my_definitions`. If you want to use those implementations in other crates,
however, you must also add a dependency to both `my_impls` and `my_definitions`
(or you could re-export the definitions in `my_impls`).

You may only implement for types/traits actually defined in `my_definitions`
itself. Re-exports don't count.

The dependency *must* be a path dependency to another crate defined in the same
workspace. Cargo will not allow you specify `impl = True` for other crates.

> Publishing packages with `impl = True` to crates.io is not allowed.

Reviewer: ```suggestion
Publishing packages with any `impl = True` dependencies to crates.io is not allowed.
```

# Guide-level explanation (rustc)

In current Rust, when you specify in your build system that a crate depends on
another, this is a "use" dependency - the dependee can use definitions in the
dependant.

This RFC defines a new dependency relationship: the "impl" dependency. The
dependee can provide implementations for types and traits defined in the
dependant.

We extend the `rustc` `--extern` flag with an additional `impl` flag: `--extern
impl:dbtrait=dbtrait.rlib`. This means that the crate currently being compiled
with this option, is allowed to implement traits and methods for types defined
in the `dbtrait` crate (the *defining* crate).

For example:
```rust
//! Defining crate
trait DBTrait {}
```
```rust
//! Implementation crate, also depends on the `mysql` crate.
use mysql::MySql;

// Allowed because we're depending on `dbtrait` with `impl`
impl dbtrait::DBTrait for MySql {}
```

In other words, the implementing crate is considered the same as the defining
crate as far as the orphan rule goes. Note that you may only implement traits
and types for crates directly defined in the defining crate &mdash; you cannot

> implement re-exported types.

Reviewer: Not sure entirely what this was supposed to mean (maybe there's a typo in there?) but, I might suggest something like:

> Note that this only applies to items directly defined in defining crate, and not to any items defined in transitive deps, regardless of whether those transitive deps are impl deps or not.

The defining and implementing crates are not the same in other respects. They
still have distinct crate names, defining their own crate-level namespaces. For
example, if they both defined a type `Foo`, it would be two definitions
`defining::Foo` and `implementing::Foo`.

There are no other constraints on the defining and implementing crates - they
can freely define new traits, types and other items, and implementations for
those types and traits. An implementing crate can be a defining crate for
another crate.

Regardless, the definitions within the defining crate, the implementing crate,
and between the implementing crate and defining crate must meet the current
coherence requirements. Also if there are multiple implementation crates in
scope in the dependency graph, they must also be coherent.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

[Coherence](https://rust-lang.github.io/chalk/book/clauses/coherence.html) is
currently defined in terms of "local" definitions - that is, items defined with

> in the current crate.

Reviewer: Hm, so I think it would help to clarify the terminology here a bit (I realize I am partially to blame for this since I may have advised you incorrectly on the terminology before). The linked document distinguishes between three concepts: 1. Coherence, which is a type of analysis result/a property that a dependency graph can have, 2. overlap checks, which are a particular algorithm to implement a component of coherence checks, and 3. the orphan rule checks, which are a separate algorithm that guarantee coherence when combined with overlap checks.

As a result, I would not describe coherence as being defined in terms of local definitions. The overlap checks specifically are though.

Author: Thanks for the clarification. I've reworded this (taking your "adoption rule" naming, since it makes sense to me now).

The details of the coherence checking logic are unchanged; the only change is
the constraints on which crates a definition and an implementation can be in.

The discussion below talks about coherence in terms of the crate dependency
relations, but the actual coherence check is at the level of individual type and
trait implementations. The crate dependency relations set an upper bound on
this, as any crate which does not appear within the relevent dependency graph
need not be considered for coherence checking.

## Changes to coherence checking

The key problem we need to solve is how to reconcile the Rust language-level
details (types and traits) with the build system view (dependency relationship
between crates). Coherence is defined in terms of type/trait definitions and
implementations, which forms its own graph. This graph is embedded in the crate
dependency graph &mdash; that is there are no edges in the type/trait

> implementation graph which are not also present in the crate dependency graph.

Reviewer: Either I'm misunderstanding the point, or this is false for blanket impls

Author: Yes, good call. This needs to be worded more carefully.

```mermaid
graph TB
  CrateC -->|dep| CrateA & CrateB
  subgraph CrateA
    A["Trait A"]
    D["Type D"]
    E["Impl A for D"] -..->|impl| A & D
  end
  subgraph CrateB
    B["Type B"]
  end
  subgraph CrateC
    C["Impl A for B"] -..->|impl| A & B
  end
```

For the discussion below, crate B is visible to crate A if it falls within its
transitive dependencies; a crate is visible to itself.

```mermaid
graph TB
  subgraph A sees only A
    c1["A"] --> c2["..."]
    c3["B"]
  end
  subgraph A sees A&B
    b1["A"] --> b2["..."] --> b3["B"]
  end
  subgraph A sees A&B
    a1["A"] --> a2["B"]
  end
```

The general rule is when compiling a crate, `rustc` must check the coherence of
a set of implementations for a type if:
- all those implementations are visible
- no other visible crate has checked them

This means the compilation of a crate must check for coherence when:
- all the definitions and implementations are within one crate
- if a crate A has an impl dependency on crate B, it must check coherency

>   between A and B

Reviewer: I don't think you mean an arbitrary crate in the dependency graph for crate A here - that seems worth clarifying

one of those DB crates in order to do `impl DBTrait for X`. It also means that
every downstream user of `DBTrait` would also transitively gain those
dependencies.

Here, `MysqlUser` ends up depending on all of `Mysql`, `MongoDB` and `RocksDB`
even though it only cares about `Mysql`.
```mermaid
graph TB
MysqlUser(MysqlUser) --> DBTrait --> Mysql & MongoDB & RocksDB
```

With this RFC, the implementation of `DBTrait` for each of the database crates
would be distributed among multiple implementation crates. Downstream users
would only need to take a dependency on the specific implemenetations they need.

In this example, `MysqlUser` depends only on `DBTrait` and `DBMysql`, and is
unaffected by the `DBMongo` and `DBRocks` implementations.
```mermaid
graph TB
Bin(MysqlUser) --> DBMysql
DBMysql & DBMongo & DBRocks -->|impl| DBTrait
DBMysql --> Mysql
DBMongo --> MongoDB
DBRocks --> RocksDB
```
# Guide-level explanation (cargo)
[guide-level-explanation]: #guide-level-explanation 

Within a workspace, you can designate an "impl" relationship between two crates.
For example, in your `my_impls/Cargo.toml` you can put:

```
[dependency]

> my_definitions = { version = "0.1", path = "../my_definitions", impl = True }

Reviewer 2: Nit: booleans are lower case in toml
https://toml.io/en/v1.0.0#boolean

```suggestion
my_definitions = { version = "0.1", path = "../my_definitions", impl = true }
```

```mermaid
graph TB
  Binary --> X & Y & Z -->|view| A & B -->|impl| C
```
`X`, `Y` and `Z` are all the first to see the implementations of `A` and `B`
with respect to `C`, so must each do their own coherency checks, even though
`Binary` only needs for this to be performed once. Fortunately these redundant
checks can be done in parallel, but it's still a waste of CPU. See below for
some discussion about optimizing this.

# Drawbacks
[drawbacks]: #drawbacks

This adds a fair amount of complexity to a delecate part of the language
semantics and to the compiler's implementation of it. This is opt-in, but one
would need to understand these semantics when using other people's packages
which use this feature.

If implemented as described here, and enabled in Cargo in an unconstrainted way,
it would enable full third-party implementations. That is, anyone could publish
a package implementing any type or trait in any other package. This would be
fine at small scalle, but at large scales it would mean that if there were two
crates with conflicting implementations, then could never appear in the same
dependency graph. That is a change to private dependencies, which wouldn't

> usually be considered compatibility breaking, would cause downstream build

Reviewer 2: This might be clearer phrasing?

```suggestion
usually be considered compatibility breaking, but can now cause downstream build
```
